lib/structural/model/has_one.rb

```diff
@@ -2,9 +2,22 @@ module Structural
   module Model
     class HasOne < Association
       def value_of(data)
-        child = data.fetch(key) { raise MissingAttributeError, key }
+        child = data.fetch(key, &default_value)
         type.new(child) unless child.nil?
       end
+
+      def default
+        valid_type_check(super)
+      end
+
+      private
+
+      def valid_type_check(v)
+        case v
+        when Hash then v
+        when Proc then valid_type_check(v.call)
+        else raise Structural::InvalidDefaultTypeError end
```

Comment:
What do you think about initialising this with `v`?  It would allow us to generate more descriptive error messages which could contain the type of `v` and `v` itself.